### PR TITLE
table: add a http api to control whether tidb check inserting 4 bytes in utf8 charset column

### PR DESCRIPTION
**Pull request by ciscoxll**
### What problem does this PR solve? 
Add a session/global variable to control whether TiDB check inserting 4 bytes in utf8 charset column.

In TiDB:
```sql
mysql> drop table if exists t2;
Query OK, 0 rows affected (0.00 sec)

mysql> create table t2(a varchar(100) charset utf8);
Query OK, 0 rows affected (0.01 sec)

mysql> insert t2 values (unhex('f09f8c80'));
ERROR 1366 (HY000): incorrect utf8 value f09f8c80(🌀) for column a
mysql> insert t2 values (unhex('F0A48BAE'));
ERROR 1366 (HY000): incorrect utf8 value f0a48bae(𤋮) for column a
```
After this pr:

```
curl -X POST -d "check_mb4_value_in_utf8=1" http://{TiDBIP}:10080/settings

Query OK, 0 rows affected (0.00 sec)

mysql> insert t2 values (unhex('f09f8c80'));
Query OK, 1 row affected (0.00 sec)

mysql> insert t2 values (unhex('F0A48BAE'));
Query OK, 1 row affected (0.00 sec)
```
### What is changed and how it works?

<del>
```sql
# Enabled
set @@session.tidb_enable_check_wrong_utf8_value = 1;
```
* `Default: session.tidb_enable_check_wrong_utf8_value = 1`

```sql
# Disable
set @@session.tidb_enable_check_wrong_utf8_value = 0;
```
</del>

```
# Enabled
curl -X POST -d "check_mb4_value_in_utf8=1" http://{TiDBIP}:10080/settings
```

```
# Disable
curl -X POST -d "check_mb4_value_in_utf8=0" http://{TiDBIP}:10080/settings
```
### Check List 

Tests 

 - Unit test

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8295)

### CODE REVIEW

File: sessionctx/variable/tidb_vars.go [hunk 1]
```diff
@@ -228,6 +228,9 @@ const (
 
 	// tidb_enable_window_function is used to control whether to enable the window function.
 	TiDBEnableWindowFunction = "tidb_enable_window_function"
+
+	// TiDBEnableCheckWrongUtf8Value is used to control whether to enable the check wrong utf8 value.
+	TiDBEnableCheckWrongUtf8Value = "tidb_enable_check_wrong_utf8_value"
```
Comment (winkyao) [inline sessionctx/variable/tidb_vars.go]: Should be `tidb_check_mb4_value_in_utf8`

**Comment by ciscoxll** [comment]
@XuHuaiyu When the charset of column is utf8 , but the mb4 character set has been inserted before.

**Comment by ciscoxll** [comment]
@crazycs520 `tidb_skip_utf8_check` is to skip verification of all UTF-8 characters, `tidb_check_mb4_value_in_utf8` just insert data without checking.

**Comment by winkyao** [comment]
@ciscoxll Please make it as a config item, don't be a session/global variable

**Comment by ciscoxll** [comment]
@winkyao PTAL.

File: sessionctx/variable/sysvar.go [hunk 2]
```diff
@@ -681,6 +681,7 @@ var defaultSysVars = []*SysVar{
 	{ScopeSession, TiDBDDLReorgPriority, "PRIORITY_LOW"},
 	{ScopeSession, TiDBForcePriority, mysql.Priority2Str[DefTiDBForcePriority]},
 	{ScopeSession, TiDBEnableRadixJoin, boolToIntStr(DefTiDBUseRadixJoin)},
+	{ScopeSession, TiDBCheckMb4ValueInUtf8, strconv.Itoa(DefEnableCheckMb4ValueInUtf8)},
```
Comment (winkyao) [inline sessionctx/variable/sysvar.go]: Remove the session variable. And add a config item

**Comment by ciscoxll** [comment]
@winkyao @crazycs520 PTAL.

**Comment by ciscoxll** [comment]
@winkyao @crazycs520 @zimulala PTAL.

File: config/config.go [hunk 3]
```diff
@@ -74,6 +74,7 @@ type Config struct {
 	Binlog              Binlog            `toml:"binlog" json:"binlog"`
 	CompatibleKillQuery bool              `toml:"compatible-kill-query" json:"compatible-kill-query"`
 	Plugin              Plugin            `toml:"plugin" json:"plugin"`
+	CheckMb4ValueInUtf8 uint32            `toml:"check-mb4-value-in-utf8" json:"check-mb4-value-in-utf8"`
```
Comment (winkyao) [inline config/config.go]: why not bool type?

File: config/config.toml.example [hunk 4]
```diff
@@ -267,3 +267,6 @@ ignore-error = false
 
 # use socket file to write binlog, for compatible with kafka version tidb-binlog.
 binlog-socket = ""
+
+# check mb4 value in utf8 is used to control whether to enable the check wrong utf8 value.
```
Comment (winkyao) [inline config/config.toml.example]: check-mb4-value-in-utf8 is used to control whether to check the mb4 characters when the charset is utf8.

File: sessionctx/variable/varsutil.go [hunk 5]
```diff
@@ -93,6 +93,8 @@ func GetSessionOnlySysVars(s *SessionVars, key string) (string, bool, error) {
 		return fmt.Sprintf("%d", s.TxnCtx.StartTS), true, nil
 	case TiDBGeneralLog:
 		return fmt.Sprintf("%d", atomic.LoadUint32(&ProcessGeneralLog)), true, nil
+	case TiDBCheckMb4ValueInUtf8:
+		return fmt.Sprintf("%d", atomic.LoadUint32(&config.GetGlobalConfig().CheckMb4ValueInUtf8)), true, nil
```
Comment (winkyao) [inline sessionctx/variable/varsutil.go]: remove the session variable. We should only use it as a config item.

**Comment by ciscoxll** [comment]
@winkyao PTAL.

**Comment by ciscoxll** [comment]
@winkyao @crazycs520 PTAL.

**Comment by crazycs520** [comment]
@winkyao @ciscoxll, Why make this variable as config item, but not be a session/global variable? Config is effective on only one tidb-server. There will some problem here:
1. If a load balance before TiDB cluster, How could you make sure this config change will take effect on all cluster? 
2. Different connection session may have different needs. How about if some session want to check_mb4_value_in_utf8, the others don't want to check.

**Comment by winkyao** [comment]
@crazycs520 this config should not be set by default，we should only disable it when the exists user using utf8 wrongly，and cannot change their codes, if we want to have back compability with them, we need a config to set it. The session/global variable is not design to solve these problems. And the global sesssion variable cannot make sense until the session reload the global variable.

**Comment by winkyao** [comment]
A config is easy to understand，and easy to maintain. But system variable is not.

**Comment by winkyao** [comment]
Maybe we should set it globally by the config item.And keep the session variable too？

**Comment by winkyao** [comment]
I added a config item already in https://github.com/pingcap/tidb/pull/9175, please fix the conflicts.

File: server/http_handler_test.go [hunk 6]
```diff
@@ -641,6 +641,20 @@ func (ts *HTTPHandlerTestSuite) TestPostSettings(c *C) {
 	c.Assert(err, IsNil)
 	c.Assert(resp.StatusCode, Equals, http.StatusOK)
 	c.Assert(atomic.LoadUint32(&variable.DDLSlowOprThreshold), Equals, uint32(200))
+
+	// test check_mb4_value_in_utf8
+	form.Set("check_mb4_value_in_utf8", "1")
+	resp, err = http.PostForm("http://127.0.0.1:10090/settings", form)
+	c.Assert(err, IsNil)
+	c.Assert(resp.StatusCode, Equals, http.StatusOK)
+	c.Assert(config.GetGlobalConfig().CheckMb4ValueInUtf8, Equals, true)
```
Comment (zimulala) [inline server/http_handler_test.go]: Add the test about insert record like "insert t2 values (unhex('f09f8c80'));"

Comment (ciscoxll) [reply server/http_handler_test.go]: @zimulala Done.

**Comment by ciscoxll** [comment]
@zimulala @winkyao PTAL.

File: server/http_handler_test.go (hunk 6 above)
Comment (ciscoxll) [reply server/http_handler_test.go]: Done.

**Comment by ciscoxll** [comment]
@zimulala @crazycs520  PTAL.

**Comment by ciscoxll** [comment]
@crazycs520 PTAL.

**Comment by ciscoxll** [comment]
@crazycs520 @zimulala PTAL.

File: sessionctx/variable/tidb_vars.go [hunk 7]
```diff
@@ -126,6 +126,9 @@ const (
 	// on: always enable table partition.
 	// off: always disable table partition.
 	TiDBEnableTablePartition = "tidb_enable_table_partition"
+
+	// TiDBEnableCheckWrongUtf8Value is used to control whether to enable the check wrong utf8 value.
+	TiDBEnableCheckWrongUtf8Value = "tidb_enable_check_wrong_utf8_value"
```
Comment (crazycs520) [inline sessionctx/variable/tidb_vars.go]: Why don't keep the same name with config `check_mb4_value_in_utf8` ?

**Comment by ciscoxll** [comment]
@crazycs520 PTAL.

**Pull request merged**